cvat/apps/engine/ddln/tasks/vls_lines/persistence/cvat.py

```diff
@@ -0,0 +1,209 @@
+class RunwayParseError(ValueError):
+    pass
+
+
+def iterate_runways(reader, reporter):
+    if not reader._frame_annotation:
+        return
+
+    lons = {}
+    lats = {}
+    for shape in reader._frame_annotation.labeled_shapes:
+        if shape.type != "rays":
+            continue
+
+        label = shape.label
+        attrs = build_attrs_dict(shape)
+        runway_id = attrs['Runway_ID']
+        if label == 'Vertical line':
+            if runway_id in lons:
+                reporter.report_duplicated_rays(runway_id, is_lon=True)
+            lons[runway_id] = (shape, attrs)
+        elif label == 'Horizontal line':
+            if runway_id in lats:
+                reporter.report_duplicated_rays(runway_id, is_lon=False)
+            lats[runway_id] = (shape, attrs)
+        else:
+            reporter.report_unknown_label(label)
+
+    for runway_id in (lats.keys() - lons.keys()):
+        reporter.report_missing_rays(runway_id, is_lon=True)
+
+    for runway_id, lon in lons.items():
+        if runway_id not in lats:
+            reporter.report_missing_rays(runway_id, is_lon=False)
+        lat = lats.get(runway_id)
+        try:
+            yield _parse_rays(lon, lat, reporter)
+        except RunwayParseError as e:
+            reporter._report(e.args[0])
+
+
+def write_runway(runway: Runway, writer):
+    lon_attrs, lat_attrs = _get_attributes(runway)
+    lon_attrs = [writer._annotations.Attribute(name=name, value=value) for name, value in lon_attrs.items()]
+    lat_attrs = [writer._annotations.Attribute(name=name, value=value) for name, value in lat_attrs.items()]
+    lon_points, lat_points = _get_points_list(runway)
+
+    lon_shape = {
+        "attributes": lon_attrs,
+        "points": lon_points,
+        "frame": writer._frame_id,
+        "group": 0,
+        "z_order": 0,
+        "occluded": False,
+        "type": "rays",
+        "label": "Vertical line",
+    }
+    lat_shape = {
+        "attributes": lat_attrs,
+        "points": lat_points,
+        "frame": writer._frame_id,
+        "group": 0,
+        "z_order": 0,
+        "occluded": False,
+        "type": "rays",
+        "label": "Horizontal line",
+    }
+    writer._annotations.add_shape(writer._annotations.LabeledShape(**lon_shape))
+    writer._annotations.add_shape(writer._annotations.LabeledShape(**lat_shape))
+
+
+def _parse_rays(lon, lat, reporter):
+    lon_shape, lon_attrs = lon
+    runway_id = lon_attrs['Runway_ID']
+    left_visible = bool(int(lon_attrs['First_line(1)']))
+    right_visible = bool(int(lon_attrs['Second_line(2)']))
+    center_visible = bool(int(lon_attrs['Third_line(3)']))
+    left, right, center = _parse_lines(lon_shape)
+    if lat:
+        lat_shape, lat_attrs = lat
+        start_visible = bool(int(lat_attrs['First_line(1)']))
+        designator_visible = bool(int(lat_attrs['Second_line(2)']))
+        end_visible = bool(int(lat_attrs['Third_line(3)']))
+        start, designator, end = _parse_lines(lat_shape)
```

Comment:
this logic looks flawed: imaging designator line is absent
then there will be only two lines, but you will parse them as `start, designator` instead of `start, end`

Reply:
ditto: this line is executed before applying visibilities, so always have 3 lines here.